### PR TITLE
Avoid unnecessary computation/locking for getStatus()

### PR DESCRIPTION
**Pull request by highker**

### CODE REVIEW

File: presto-main/src/main/java/com/facebook/presto/execution/SqlTask.java [hunk 1]
```diff
@@ -212,17 +213,32 @@ private TaskStatus createTaskStatus(TaskHolder taskHolder)
...
+        DataSize memoryReservation = new DataSize(0, BYTE);
+        if (taskHolder.getFinalTaskInfo() != null) {
+            TaskStats taskStats = taskHolder.getFinalTaskInfo().getStats();
+            queuedPartitionedDrivers = taskStats.getQueuedPartitionedDrivers();
+            runningPartitionedDrivers = taskStats.getRunningPartitionedDrivers();
+            memoryReservation = taskStats.getMemoryReservation();
+        }
+        else if (taskHolder.getTaskExecution() != null) {
+            TaskContext taskContext = taskHolder.getTaskExecution().getTaskContext();
+            queuedPartitionedDrivers = taskContext.getPipelineContexts().stream().mapToInt(context -> context.getPipelineStatus().getQueuedPartitionedDrivers()).sum();
```
Comment (dain) [inline presto-main/src/main/java/com/facebook/presto/execution/SqlTask.java]: I'd do this with a traditional `for` loop and do both sums at the same time

File: presto-main/src/main/java/com/facebook/presto/operator/ExchangeClient.java [hunk 2]
```diff
@@ -114,11 +114,14 @@ public ExchangeClient(
         this.maxBufferBytes = Long.MIN_VALUE;
     }
 
-    public synchronized ExchangeClientStatus getStatus()
```
Comment (dain) [inline presto-main/src/main/java/com/facebook/presto/operator/ExchangeClient.java]: I'm pretty sure this violates the Java memory model.  The fields like `bufferBytes` at the bottom must be read within a lock.

Maybe reorder the code so the `allClients` loop is first and then you take the synchronized lock for the rest of the code.

**Comment by highker** [comment]
Comments addressed

File: presto-main/src/main/java/com/facebook/presto/operator/ExchangeClient.java [hunk 3]
```diff
@@ -114,18 +114,23 @@ public ExchangeClient(
...
     {
-        int bufferedPages = pageBuffer.size();
-        if (bufferedPages > 0 && pageBuffer.peekLast() == NO_MORE_PAGES) {
-            bufferedPages--;
-        }
-
-        ImmutableList.Builder<PageBufferClientStatus> exchangeStatus = ImmutableList.builder();
+        // The stats created by this method serves for the purpose of debugging.
+        // It does not guarantee a consistent view of the clients.
+        // The reason is to reduce the lock contention.
```
Comment (haozhun) [inline presto-main/src/main/java/com/facebook/presto/operator/ExchangeClient.java]:
```
        // The stats created by this method is only for diagnostics.
        // It does not guarantee a consistent view between different exchange clients.
        // Guaranteeing a consistent view introduces significant lock contention.

File: presto-main/src/main/java/com/facebook/presto/operator/ExchangeClient.java [hunk 4]
```diff
@@ -114,18 +114,23 @@ public ExchangeClient(
...
+        ImmutableList.Builder<PageBufferClientStatus> pageBufferClientStatusBuilder = ImmutableList.builder();
         for (HttpPageBufferClient client : allClients.values()) {
-            exchangeStatus.add(client.getStatus());
+            pageBufferClientStatusBuilder.add(client.getStatus());
+        }
+        ImmutableList<PageBufferClientStatus> pageBufferClientStatus = pageBufferClientStatusBuilder.build();
```
Comment (haozhun) [inline presto-main/src/main/java/com/facebook/presto/operator/ExchangeClient.java]: `List`

File: presto-main/src/main/java/com/facebook/presto/operator/DriverContext.java [hunk 5]
```diff
@@ -354,6 +354,16 @@ public CounterStat getOutputPositions()
         }
     }
 
+    public boolean executionStarted()
```
Comment (haozhun) [inline presto-main/src/main/java/com/facebook/presto/operator/DriverContext.java]: isExecutionStarted

File: presto-main/src/main/java/com/facebook/presto/operator/PipelineContext.java [hunk 6]
```diff
@@ -343,6 +343,11 @@ public CounterStat getOutputPositions()
         return stat;
     }
 
+    public PipelineStatus getPipelineStatus()
+    {
+        return new PipelineStatus(ImmutableList.copyOf(this.drivers));
```
Comment (haozhun) [inline presto-main/src/main/java/com/facebook/presto/operator/PipelineContext.java]: Make the `PipelineStatus` constructor as dumb as possible. Put logic here instead.

**Comment by highker** [comment]
Comments addressed; thanks for the review.

File: presto-main/src/main/java/com/facebook/presto/operator/PipelineStatus.java [hunk 7]
```diff
@@ -0,0 +1,60 @@
...
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+package com.facebook.presto.operator;
+
+import javax.annotation.concurrent.Immutable;
+
+@Immutable
+public class PipelineStatus
```
Comment (nezihyigitbasi) [inline presto-main/src/main/java/com/facebook/presto/operator/PipelineStatus.java]: `public final`

File: presto-main/src/main/java/com/facebook/presto/operator/PipelineContext.java [hunk 8]
```diff
@@ -513,4 +498,32 @@ else if (driverStats.isFullyBlocked()) {
 
         return map.replace(key, oldValue, newValue);
     }
+
+    private static PipelineStatus getPipelineStatus(List<DriverContext> driverContexts)
```
Comment (nezihyigitbasi) [inline presto-main/src/main/java/com/facebook/presto/operator/PipelineContext.java]: you don't need to make this `static`

**Pull request merged**